### PR TITLE
Add clientEndpointType to interface

### PR DESCRIPTION
**Pull request by cqnguy23**
# Description

1. Support to generate MQTT JWT Token for Azure WebPubSub. Essentially we need to identify if clients want to establish a default WebPubSub connection or an MQTT connection, then we will generate the Client Connection URL for them accordingly.

- Example Client URL for:
  - Default Connection: `wss://exampleHost.com/client/hubs/exampleHub`
  - MQTT Connection: `wss://exampleHost.com/clients/mqtt/hubs/exampleHub`
 
2. Upgrade API version to '2024-01-01'
3. Add tests to get MQTT Token

### CODE REVIEW

File: sdk/webpubsub/azure-messaging-webpubsub/src/main/java/com/azure/messaging/webpubsub/models/GetClientAccessTokenOptions.java [hunk 1]
```diff
@@ -20,7 +20,7 @@ public final class GetClientAccessTokenOptions {
     private String userId;
     private List<String> roles;
     private List<String> groups;
-
+    private ClientEndpointType clientEndpointType;
```
Comment (Y-Sindo) [inline sdk/webpubsub/azure-messaging-webpubsub/src/main/java/com/azure/messaging/webpubsub/models/GetClientAccessTokenOptions.java:23]: Do we need to specify default value ?

File: sdk/webpubsub/azure-messaging-webpubsub/src/main/java/com/azure/messaging/webpubsub/models/ClientEndpointType.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+import com.fasterxml.jackson.annotation.JsonCreator;
+
+/**
+ * Defines values for ClientEndpointType
+ */
+public enum ClientEndpointType {
+    /** Default WebPubSub Client Endpoint. E.g: <code>wss://exampleHost.com/client/hubs/exampleHub</code> */
+    DEFAULT("default"),
+
+    /** MQTT Client Endpoint. E.g: <code>wss://exampleHost.com/client/mqtt/hubs/exampleHub</code> */
```
Comment (Y-Sindo) [inline sdk/webpubsub/azure-messaging-webpubsub/src/main/java/com/azure/messaging/webpubsub/models/ClientEndpointType.java:12]: It should be `/clients` (an 's' prefixed). You may have already noticed that it's a different pattern with the `/client/hubs/hub`

File: sdk/webpubsub/azure-messaging-webpubsub/src/main/java/com/azure/messaging/webpubsub/models/ClientEndpointType.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    ClientEndpointType(String value) { this.value = value; }
+
+    /**
+     * Parses a serialized value to a ClientEndpointType instance.
+     *
+     * @param value the serialized value to parse.
+     * @return the parsed ClientEndpointType object, or null if unable to parse.
+     */
+    @JsonCreator
+    public static ClientEndpointType fromString(String value) {
```
Comment (Y-Sindo) [inline sdk/webpubsub/azure-messaging-webpubsub/src/main/java/com/azure/messaging/webpubsub/models/ClientEndpointType.java:27]: Why do we need serialization method here?

Comment (cqnguy23) [reply sdk/webpubsub/azure-messaging-webpubsub/src/main/java/com/azure/messaging/webpubsub/models/ClientEndpointType.java:27]: We don't currently need it. But it seems that most enum classes in the repos define their own serialize and de-serialize methods, so this is just to follow conventions

**Comment by weidongxu-microsoft** [comment]
Please run a regen via swagger/readme.md

At least, the ServiceVersion class would need a new api-version
https://github.com/Azure/azure-sdk-for-java/blob/4ca954f1abf9165b4589f9f7f7cc351e9b92160b/sdk/webpubsub/azure-messaging-webpubsub/src/main/java/com/azure/messaging/webpubsub/WebPubSubServiceVersion.java

File: sdk/webpubsub/azure-messaging-webpubsub/src/main/java/com/azure/messaging/webpubsub/WebPubSubServiceAsyncClient.java [hunk 4]
```diff
@@ -20,10 +20,7 @@
 import com.azure.core.util.FluxUtil;
 import com.azure.messaging.webpubsub.implementation.WebPubSubUtil;
 import com.azure.messaging.webpubsub.implementation.WebPubSubsImpl;
-import com.azure.messaging.webpubsub.models.GetClientAccessTokenOptions;
-import com.azure.messaging.webpubsub.models.WebPubSubClientAccessToken;
-import com.azure.messaging.webpubsub.models.WebPubSubContentType;
-import com.azure.messaging.webpubsub.models.WebPubSubPermission;
+import com.azure.messaging.webpubsub.models.*;
```
Comment (weidongxu-microsoft) [inline sdk/webpubsub/azure-messaging-webpubsub/src/main/java/com/azure/messaging/webpubsub/WebPubSubServiceAsyncClient.java]: don't do `import .*`, fix your IDE

File: sdk/webpubsub/azure-messaging-webpubsub/src/main/java/com/azure/messaging/webpubsub/WebPubSubServiceAsyncClient.java [hunk 5]
```diff
@@ -55,18 +52,21 @@ public final class WebPubSubServiceAsyncClient {
      */
     @ServiceMethod(returns = ReturnType.SINGLE)
     public Mono<WebPubSubClientAccessToken> getClientAccessToken(GetClientAccessTokenOptions options) {
+        final ClientEndpointType clientEndpointType = options.getClientEndpointType();
+        final String path = clientEndpointType.equals(ClientEndpointType.MQTT)
```
Comment (weidongxu-microsoft) [inline sdk/webpubsub/azure-messaging-webpubsub/src/main/java/com/azure/messaging/webpubsub/WebPubSubServiceAsyncClient.java:60]: ClientEndpointType seems not committed to PR?

Comment (weidongxu-microsoft) [reply sdk/webpubsub/azure-messaging-webpubsub/src/main/java/com/azure/messaging/webpubsub/WebPubSubServiceAsyncClient.java:60]: Also I didn't see change to GetClientAccessTokenOptions.java

**Review by weidongxu-microsoft — COMMENTED** [review]
You will also need to update CHANGELOG.md

File: sdk/webpubsub/azure-messaging-webpubsub/src/main/java/com/azure/messaging/webpubsub/WebPubSubServiceAsyncClient.java (hunk 4 above)
Comment (cqnguy23) [reply sdk/webpubsub/azure-messaging-webpubsub/src/main/java/com/azure/messaging/webpubsub/WebPubSubServiceAsyncClient.java]: this seems to be on an outdated commit, can you recheck?

File: sdk/webpubsub/azure-messaging-webpubsub/src/main/java/com/azure/messaging/webpubsub/WebPubSubServiceAsyncClient.java (hunk 5 above)
Comment (cqnguy23) [reply sdk/webpubsub/azure-messaging-webpubsub/src/main/java/com/azure/messaging/webpubsub/WebPubSubServiceAsyncClient.java:60]: Oh I'm seeing both changes are included in the PR already. Can you double check? 
https://github.com/Azure/azure-sdk-for-java/pull/40784/files#diff-378f75168506064970a2cbfd5ad1647468e502cba50e29073a0e099db491ef7c
https://github.com/Azure/azure-sdk-for-java/pull/40784/files#diff-4a2e7e8521c6c5d25ae24a7f3d8b78066f189dbd70d04c8e1344fc737f805c71

File: sdk/webpubsub/azure-messaging-webpubsub/src/test/java/com/azure/messaging/webpubsub/WebPubSubServiceClientTests.java [hunk 6]
```diff
@@ -14,10 +14,7 @@
 import com.azure.core.test.annotation.DoNotRecord;
 import com.azure.core.util.BinaryData;
 import com.azure.identity.DefaultAzureCredentialBuilder;
-import com.azure.messaging.webpubsub.models.GetClientAccessTokenOptions;
-import com.azure.messaging.webpubsub.models.WebPubSubClientAccessToken;
-import com.azure.messaging.webpubsub.models.WebPubSubContentType;
-import com.azure.messaging.webpubsub.models.WebPubSubPermission;
+import com.azure.messaging.webpubsub.models.*;
```
Comment (weidongxu-microsoft) [inline sdk/webpubsub/azure-messaging-webpubsub/src/test/java/com/azure/messaging/webpubsub/WebPubSubServiceClientTests.java:17]: and this?

File: sdk/webpubsub/azure-messaging-webpubsub/src/main/java/com/azure/messaging/webpubsub/models/ClientEndpointType.java [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+// Licensed under the MIT License.
+
+package com.azure.messaging.webpubsub.models;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+
+/**
+ * Defines values for ClientEndpointType
+ */
+public enum ClientEndpointType {
```
Comment (weidongxu-microsoft) [inline sdk/webpubsub/azure-messaging-webpubsub/src/main/java/com/azure/messaging/webpubsub/models/ClientEndpointType.java:11]: `enum` here means no value can be added, ever (new api-version included). Make sure your intention is such.

File: sdk/webpubsub/azure-messaging-webpubsub/src/main/java/com/azure/messaging/webpubsub/models/ClientEndpointType.java [hunk 8]
```diff
@@ -0,0 +1,52 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.messaging.webpubsub.models;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
```
Comment (weidongxu-microsoft) [inline sdk/webpubsub/azure-messaging-webpubsub/src/main/java/com/azure/messaging/webpubsub/models/ClientEndpointType.java:6]: No `com.fasterxml.jackson` import.

File: sdk/webpubsub/azure-messaging-webpubsub/src/test/java/com/azure/messaging/webpubsub/WebPubSubServiceClientTests.java (hunk 6 above)
Comment (cqnguy23) [reply sdk/webpubsub/azure-messaging-webpubsub/src/test/java/com/azure/messaging/webpubsub/WebPubSubServiceClientTests.java:17]: my bad, I'll change it back

File: sdk/webpubsub/azure-messaging-webpubsub/CHANGELOG.md [hunk 9]
```diff
@@ -1,5 +1,15 @@
...
+  when generating token. This option can be used to generate token and client connection URL for a specific client endpoint type, such as `Default` or `MQTT`.
+
+#### Dependency Updates
+
+- Updated `api-version` to `2024-01-01`.
+
+
 ## 1.3.0-beta.1 (Unreleased)
 
 - Migrated serialization to `azure-json` which offers implementation agnostic serialization, providing support for
```
Comment (weidongxu-microsoft) [inline sdk/webpubsub/azure-messaging-webpubsub/CHANGELOG.md:15]: Please merge the section with 1.3.0-beta.1
(1.3.0-beta.1 is not released, so the changes there also applies to 1.3.0)

Also please add a section of below (the exact lib version depends on your release date -- if you follow the July release, you will need to wait a few days for new azure-core)

```
### Other Changes
#### Dependency Updates
- Upgraded `azure-core` from `1.47.0` to version `1.48.0`.
- Upgraded `azure-core-http-netty` from `1.14.1` to version `1.14.2`.
```

File: sdk/webpubsub/azure-messaging-webpubsub/src/test/java/com/azure/messaging/webpubsub/WebPubSubServiceClientTests.java [hunk 10]
```diff
@@ -28,9 +30,7 @@
 import java.nio.charset.StandardCharsets;
 import java.text.ParseException;
 
-import static org.junit.jupiter.api.Assertions.assertEquals;
-import static org.junit.jupiter.api.Assertions.assertNotNull;
-import static org.junit.jupiter.api.Assumptions.assumeTrue;
+import static org.junit.jupiter.api.Assertions.*;
```
Comment (weidongxu-microsoft) [inline sdk/webpubsub/azure-messaging-webpubsub/src/test/java/com/azure/messaging/webpubsub/WebPubSubServiceClientTests.java:33]: Don't do `import *`

File: sdk/webpubsub/azure-messaging-webpubsub/CHANGELOG.md (hunk 9 above)
Comment (cqnguy23) [reply sdk/webpubsub/azure-messaging-webpubsub/CHANGELOG.md:15]: So for the Dependency Updates, we just add a temporary section like what you suggested, then we'll change it to the correct lib version when we actually release right?

Comment (weidongxu-microsoft) [reply sdk/webpubsub/azure-messaging-webpubsub/CHANGELOG.md]: What's your scheduled release date?

If you don't have a date yet, I'd suggest you put e.g. `1.3.0 (Unreleased)`, without the "Dependency Updates" section.
You will use the Prepare-Release script to update the CHANGELOG upon your release (when you can add the "Dependency Updates" section with exact version).

File: sdk/webpubsub/azure-messaging-webpubsub/src/main/java/com/azure/messaging/webpubsub/WebPubSubServiceClient.java [hunk 11]
```diff
@@ -39,7 +42,7 @@ public final class WebPubSubServiceClient {
      * @param serviceClient the service client implementation.
      */
     WebPubSubServiceClient(WebPubSubsImpl serviceClient, String hub, String endpoint,
-        AzureKeyCredential keyCredential) {
+                           AzureKeyCredential keyCredential) {
```
Comment (haolingdong-msft) [inline sdk/webpubsub/azure-messaging-webpubsub/src/main/java/com/azure/messaging/webpubsub/WebPubSubServiceClient.java:45]: Better to avoid these unnecessary formatting.

Comment (cqnguy23) [reply sdk/webpubsub/azure-messaging-webpubsub/src/main/java/com/azure/messaging/webpubsub/WebPubSubServiceClient.java:45]: These are actually automatically generated when I run Autorest. Should I undo it instead?

File: sdk/webpubsub/azure-messaging-webpubsub/src/main/java/com/azure/messaging/webpubsub/implementation/models/AddToGroupsRequest.java [hunk 12]
```diff
@@ -0,0 +1,111 @@
+package com.azure.messaging.webpubsub.implementation.models;
```
Comment (weidongxu-microsoft) [inline sdk/webpubsub/azure-messaging-webpubsub/src/main/java/com/azure/messaging/webpubsub/implementation/models/AddToGroupsRequest.java:1]: copyright

File: sdk/webpubsub/azure-messaging-webpubsub/src/main/java/com/azure/messaging/webpubsub/WebPubSubServiceClient.java [hunk 13]
```diff
@@ -344,7 +347,8 @@ public Response<Void> addConnectionToGroupWithResponse(String group, String conn
      *
      * @param hub Target hub name, which should start with alphabetic characters and only contain alpha-numeric
      * characters or underscore.
-     * @param groupsToAdd Target groups and connection filter.
+     * @param groups Target group names
```
Comment (weidongxu-microsoft) [inline sdk/webpubsub/azure-messaging-webpubsub/src/main/java/com/azure/messaging/webpubsub/WebPubSubServiceClient.java:350]: I see you send `null` in test. Please clarify this in Javadoc (i.e., what would be the effect of the API, if this param is null).

Same to the Async client.

File: sdk/webpubsub/azure-messaging-webpubsub/src/main/java/com/azure/messaging/webpubsub/WebPubSubServiceClient.java [hunk 14]
```diff
@@ -353,9 +357,14 @@ public Response<Void> addConnectionToGroupWithResponse(String group, String conn
      * @throws ResourceModifiedException thrown if the request is rejected by server on status code 409.
      */
     @ServiceMethod(returns = ReturnType.SINGLE)
-    public Response<Void> addConnectionsToGroupsWithResponse(String hub, BinaryData groupsToAdd,
+    public Response<Void> addConnectionsToGroupsWithResponse(String hub, List<String> groups, String filter,
```
Comment (weidongxu-microsoft) [inline sdk/webpubsub/azure-messaging-webpubsub/src/main/java/com/azure/messaging/webpubsub/WebPubSubServiceClient.java:360]: Should this API be `void addConnectionsToGroups(String hub, List<String> groups, String filter)`?

Comment (cqnguy23) [reply sdk/webpubsub/azure-messaging-webpubsub/src/main/java/com/azure/messaging/webpubsub/WebPubSubServiceClient.java:360]: Just checked with Liangying, for this we'll keep the current method signature to follow the current convention (other methods also follow this pattern as well)
cc @vicancy

Comment (weidongxu-microsoft) [reply sdk/webpubsub/azure-messaging-webpubsub/src/main/java/com/azure/messaging/webpubsub/WebPubSubServiceClient.java:360]: Please see this API and their differences. Generally speaking, the `WithResponse` API is not recommended to be updated. You can hide it, but not recommend to do such modification on it.
https://github.com/Azure/azure-sdk-for-java/blob/deef3da0ab9b030bcf28c8b941f45764380b6e78/sdk/webpubsub/azure-messaging-webpubsub/src/main/java/com/azure/messaging/webpubsub/WebPubSubServiceClient.java#L121-L145

Loop @srikanta

Comment (cqnguy23) [reply sdk/webpubsub/azure-messaging-webpubsub/src/main/java/com/azure/messaging/webpubsub/WebPubSubServiceClient.java:360]: okay understood, I'll undo the changes on the WithResponse API method and wrap it with a void method then

Comment (cqnguy23) [reply sdk/webpubsub/azure-messaging-webpubsub/src/main/java/com/azure/messaging/webpubsub/WebPubSubServiceClient.java]: @weidongxu-microsoft just pushed new changes. Please let me know if anything else needs to improve

Comment (weidongxu-microsoft) [reply sdk/webpubsub/azure-messaging-webpubsub/src/main/java/com/azure/messaging/webpubsub/WebPubSubServiceClient.java]: `addConnectionsToGroups`

no `requestOptions` param

Same apply to the Async client.

Comment (cqnguy23) [reply sdk/webpubsub/azure-messaging-webpubsub/src/main/java/com/azure/messaging/webpubsub/WebPubSubServiceClient.java]: @weidongxu-microsoft done. Can you please recheck?

Comment (weidongxu-microsoft) [reply sdk/webpubsub/azure-messaging-webpubsub/src/main/java/com/azure/messaging/webpubsub/WebPubSubServiceClient.java]: The public method is `addConnectionsToGroups`. No `WithResponse`.

File: sdk/webpubsub/azure-messaging-webpubsub/CHANGELOG.md [hunk 15]
```diff
@@ -2,6 +2,8 @@
 
 ## 1.3.0-beta.1 (Unreleased)
 
+- Added a `clientEndpointType` option to `GenerateClientTokenOptions` to specify the type of client endpoint
+  when generating token. This option can be used to generate token and client connection URL for a specific client endpoint type, such as `Default` or `MQTT`.
 - Migrated serialization to `azure-json` which offers implementation agnostic serialization, providing support for
   more serialization frameworks than just Jackson.
```
Comment (weidongxu-microsoft) [inline sdk/webpubsub/azure-messaging-webpubsub/CHANGELOG.md:8]: Put these in `### Features Added` section

Also, do you need to mention the added API addConnectionsToGroups?

**Review by weidongxu-microsoft — COMMENTED** [review]
LGTM

What do you think @vicancy @srnagar @haolingdong-msft

File: sdk/webpubsub/azure-messaging-webpubsub/CHANGELOG.md (hunk 15 above)
Comment (cqnguy23) [reply sdk/webpubsub/azure-messaging-webpubsub/CHANGELOG.md:8]: Yeah I need to do that, will update it now

**Review by vicancy — APPROVED** [review]

**Review by weidongxu-microsoft — APPROVED** [review]

**Review by haolingdong-msft — COMMENTED** [review]
LGTM.
